Add calibration to CQL as in CalQL paper arXiv:2303.05479

BFAnas opened the pull request:
- [X] I have marked all applicable categories:
    + [ ] exception-raising fix
    + [ ] algorithm implementation fix
    + [ ] documentation modification
    + [X] new feature
- [X] I have reformatted the code using `make format` (**required**)
- [X] I have checked the code using `make commit-checks` (**required**)
- [X] If applicable, I have mentioned the relevant/related issue(s)
- [X] If applicable, I have listed every items in this Pull Request below

MischaPanch commented:
If possible, pls wait until #912 is merged, as it will certainly introduce conflicts. It is now ready to be merged (CI is green), so I hope it will be quick

BFAnas commented:
Yes sure no problem. I close this one for now until after your merge.

MischaPanch commented:
Thank you, appreciate it!

MischaPanch commented:
You can go for it now ;)

BFAnas commented:
Awsome, congrats on the big Merge!

MischaPanch commented:
@BFAnas I will try review this in the near future. Could you pls resolve the conflicts with master? We now have a new built and the newest formatting (for py3.11)

MischaPanch commented on examples/offline/d4rl_cql.py:
```diff
@@ -73,6 +74,33 @@ def get_args():
     return parser.parse_args()
 
 
+def add_returns(buffer: ReplayBuffer, gamma: float = 0.99) -> ReplayBuffer:
```
There is a function in `policy.base` called `_gae_return` which is used for calculating returns in `BasePolicy.compute_episodic_return`. Why did you have to add your own one? Is it different from the already existing functionality?

BFAnas replied:
`_gae_return` doesn't compute returns but rather advantages, and `BasePolicy.compute_episodic_return` acts on `Batch` rather than the `ReplayBuffer`, it can be used in 'BasePolicy.preprocess_fn' however having to preprocess each batch separately slows the execution considerably.

MischaPanch replied:
Adding the value to the gae gives you the return. See the docstring there

MischaPanch commented on examples/offline/d4rl_cql.py:
```diff
@@ -202,6 +231,8 @@ def watch():
 
     if not args.watch:
         replay_buffer = load_buffer_d4rl(args.expert_data_task)
+        if args.calibrated:
```
Should this option be part of the CQL implementation instead of the example?

BFAnas replied:
It's better to preprocess the whole buffer at once and one time, than to preprocess batches as they arrive in the CQL implementation.

MischaPanch replied:
See below, you can do that in `process_fn`

MischaPanch replied:
Closing, this is discussed in other threads

MischaPanch commented on test/offline/test_cql.py:
```diff
@@ -68,15 +68,43 @@ def get_args():
 
 
-def test_cql(args=get_args()):
+def add_returns(buffer: ReplayBuffer, gamma: float = 0.99) -> ReplayBuffer:
```
this duplication is a sign of something going wrong, see my comments above

MischaPanch commented on tianshou/policy/imitation/cql.py:
```diff
@@ -251,6 +255,23 @@ def learn(self, batch: RolloutBatchProtocol, *args: Any, **kwargs: Any) -> dict[
         ]:
             value.reshape(batch_size, self.num_repeat_actions, 1)
 
+        if self.calibrated:
```
Note that with the current design, the user has to know to add returns to the batch under the exactly right field name, which means this code is fragile.

Instead, when `calibrated=True`, the returns should be added to the batch inside the policy class itself. Make sure to use the right `BatchProtocol` for expressing that, if needed (or make a new one, if there is no suitable).

BFAnas replied:
I agree that this is a fragile implementation, however the solution you are proposing comes at a high performance cost. Maybe adding some checks is the way to go for now?

MischaPanch replied:
Why would there be a performance cost? You can add new values to the batch in `process_fn`, which takes the whole buffer. Am I missing something?

Adding checks is not the way to go here, we can't expect a Policy to only work if a user creates some external function that carefully processes the buffer

BFAnas replied:
`process_fn` is executed at every learning step that's where the performance cost comes from, compared to doing it only one time before starting training. I think this comes from the fundamental difference between offline and online RL, so maybe better to include a preprocess buffer step to the OfflineTrainer where this kind of one time modifications that are possible in offline RL are possible.

MischaPanch replied:
Ah, I see. I want to adjust the control flow of update->preprocess->learn in the near future, in part to avoid such problems. Memoization would also help here... It can't be solved in the trainer, as the trainer doesn't know about the `calibrated` arg.

I suggest the following: 
1. Add a new parameter to CQLPolicy called `offline_mode: bool = True`. Document that with this, things like `process_fn` will only be called once.
2. Add hidden fields a la `_buffer_was_processed = False`
3. Use this field to make sure `process_fn` is called only once during training. You should also be able to reset the fields, maybe adding a `reset` method to the policy

It's not particularly clean, but should do the job until the general structure has been improved

BFAnas replied:
The execution of `process_fn` is outside the policy implementation, it is called in other places such as `collector`, so unless I'm missing something, I don't think it's advisable nor possible to specify how many times that it should be called. My idea is to add a `preprocess_buffer` to the policy, that if it exists in the policy implementation it will be called once in the OfflineTrianer.

MischaPanch replied:
Ok, that's even better! Then it should be added to `BasePolicy`, where by default it just returns the unmodified buffer, and overwritten with `@overwrite` in CQLPolicy. The BaseTrainer can be adjusted to always call this function. Could you pls also add this logic to the Trainer's documentation?

MischaPanch commented on tianshou/policy/imitation/cql.py:
```diff
@@ -45,6 +45,7 @@ class CQLPolicy(SACPolicy):
     :param float alpha_min: lower bound for clipping cql_alpha. Default to 0.0.
     :param float alpha_max: upper bound for clipping cql_alpha. Default to 1e6.
     :param float clip_grad: clip_grad for updating critic network. Default to 1.0.
+    :param calibrated: calibrate Q-values as in CalQL paper arXiv:2303.05479.
```
Just referencing the paper is not enough, pls provide a short explanation of what this does and when it makes sense to use this

BFAnas replied:
Added information about use case

MischaPanch commented on tianshou/policy/imitation/cql.py:
```diff
@@ -251,6 +255,23 @@ def learn(self, batch: RolloutBatchProtocol, *args: Any, **kwargs: Any) -> dict[
+                )
+                .view(-1, 1)
+            )
+            random_value1 = torch.max(random_value1, returns)
```
should be done in a for loop that modifies the arrays inplace. You can define a new var that holds all these arrays, and also use it in the loop above

BFAnas replied:
You mean something like this? do you have an improvement?

```
            values = [random_value1, random_value2, current_pi_value1, current_pi_value2, next_pi_value1, next_pi_value2]
            for i in range(len(values)):
                values[i] = torch.max(values[i], returns)

            random_value1, random_value2, current_pi_value1, current_pi_value2, next_pi_value1, next_pi_value2 = tuple(values)
```

MischaPanch replied:
rather 

`torch.max(values, returns, out=values)`

MischaPanch replied:
If you get issues with arrays and `out`, you can also do 

`values[:] = torch,max(values, returns)`

but my first proposal is more elegant. Python for loops should be avoided as much as possible

BFAnas replied:
max() received an invalid combination of arguments - got (list, Tensor), but expected one of:
 * (Tensor input, *, Tensor out)
 * (Tensor input, Tensor other, *, Tensor out)
 * (Tensor input, int dim, bool keepdim, *, tuple of Tensors out)
 * (Tensor input, name dim, bool keepdim, *, tuple of Tensors out)
 
 
This  `values[:] = torch,max(values, returns)` is not working, and I don't think having a for loop really makes the code that elegant since we have to reassign the variables after the for loop. Closing and leaving the code as it is.

MischaPanch replied:
We misunderstood each other, what I meant was

```
for t in tensors:
   torch.max(t, returns, out=t)
```

It's easy to make mistakes when many names are repeated, and it also makes it a bit harder to read. I can push this myself later if you want to

BFAnas replied:
Yes you can go ahead and give it a try, but keep in mind that `torch.max(t, returns, out=t)` doesn't work with tensors that need differentiation.

MischaPanch commented on test/offline/test_cql.py:
```diff
@@ -68,15 +68,43 @@ def get_args():
     if os.path.exists(args.load_buffer_name) and os.path.isfile(args.load_buffer_name):
         if args.load_buffer_name.endswith(".hdf5"):
-            buffer = VectorReplayBuffer.load_hdf5(args.load_buffer_name)
+            buffer = ReplayBuffer.load_hdf5(args.load_buffer_name)
```
Why was that needed?

BFAnas replied:
It's the same in fact, `VectorReplayBuffer` just inherits `load_hdf5` from `ReplayBuffer`. And having `VectorReplayBuffer.load_hdf5' is kind of misleading that buffer should be a vector buffer.

MischaPanch replied:
It's not the same, since load_hdf5 is a classmethod and despite being inherited, it would instantiate a VectorReplayBuffer and not a ReplayBuffer. The type is not annotated sufficiently well to express it, I will open a mini PR now to fix it

BFAnas replied:
I think for offline we don't need `VectorReplayBuffer` since (at least as understand it) it is meant to work with `VectorEnv` but in offline we don't have an `Env`. But anyway for that code both work fine without a problem.

MischaPanch replied:
Resolving. FYI, the type issue is fixed with #942

MischaPanch replied:
Unresolved this, pls change back to VectorReplayBuffer. Even if ReplayBuffer makes sense here, should happen in a separate PR

MischaPanch commented on test/offline/test_cql.py:
```diff
@@ -227,3 +256,5 @@ def watch():
 if __name__ == "__main__":
     test_cql()
+    # test calibrated cql
+    test_cql(calibrated=True)
```
Does it really test that the new functionality is performing what it's supposed to do?

MischaPanch commented on test/offline/test_cql.py:
```diff
@@ -68,15 +68,43 @@ def get_args():
     return parser.parse_known_args()[0]
 
 
-def test_cql(args=get_args()):
```
From now on, more unit tests should be added to tianshou - just relying on integration tests can lead to serious bugs. Pls make sure to add tests that verify that the extension is doing what it's supposed to do.

BFAnas replied:
Yes I agree, however not sure how can we test if an RL algorithm is really doing what it is intending to do. One way would be to be to save some deterministic values that have been identified as the values that should be reproduced if everything is working ok, and then reproduce them in test. However, this is practically impossible given that any change in the software stack can affect those values and trigger a false negative. 
Do you have some suggestions on how to do it?

MischaPanch replied:
Testing performance is indeed difficult, but unit testing should be ok. The fact that it's difficult now is a reflection of imperfect design, like the very long implementation of `learn`. If you create a new method that returns what is now called `catq1/2`, you could test then when `calibrated=True`, the max is <= the max of the returns.

You can also test that the processed buffer has the correct field if `calibrated=True` after preprocess of the buffer, as discussed above.

Adding a second integration test with `calibration=True` serves relatively little purpose, but increases the runtime by a lot. Over time, the tests will be adjusted to be more useful and faster - this will require breaking up long methods in smaller ones. It's a lot of work, but it's very important for the overall future of tianshou and the developer experiences. Let's start this process now, ok? At least, new code shouldn't add to the current problems of tianshou :) 

Thank you again for your contribution and patience, I appreciate it!

MischaPanch replied:
Regarding performance - monkey patching is the way to go here, it allows mimicking performance outputs but testing everything else. But that goes beyond the scope of this PR

BFAnas replied:
Thanks for your thorough reviews and your contributions as well :)
For the testing, I think what you are proposing boils down to testing if the max operation is done correctly and if the returns were calculated correctly.
I don't think testing the correctness of RL algorithm implementations can be automated in this way (or at all unless more sophisticated AI is invented).

BFAnas replied:
I tested those manually to check if the implementation is correct, and I don't think it is useful to test it every time. Until a better system is developed I don't run that running tests that verify those two things makes sense when all the other parts of the algorithm are not tested and as I mentioned I believe that it's not something that can be automatized. 
I think you are right about avoiding to run CQL two times, I'll set calibrated to True as default and just run CQL once. In my testing the calibrated version performs better across the board so it might as well be the default.

MischaPanch replied:
It's not only about being correct now but also about future developers who want to know that they haven't broken stuff by making changes. The latter is very important. 

Apart from that, the absence of unit tests in tianshou has led to untestable (i.e. not will designed) code, as you have encountered now. At least I would want new contributions to make the overall code quality slightly better, not slightly worse.

But I understand if you don't want to invest additional time on it now. We can merge this first and I will add a unit test for this in a follow up, and assign you as reviewer, if you prefer that

BFAnas replied:
For me it's just in principle it doesn't make sense to have automated tests for RL algos otherwise we should test every line of the code which doesn't make sense. I believe RL algos should be tested manually by looking at losses and other metrics not this way. If I'm convinced otherwise I would happily do it.

MischaPanch replied:
I don't think I can say something convincing here, so let me do a follow-up PR. Maybe the way I write it will give you a better feeling for what I mean

BFAnas commented:
@MischaPanch some mypy checks fail that are not in my part of the code, some are with some notes do you know about them?

MischaPanch commented on examples/offline/d4rl_cql.py:
```diff
@@ -22,9 +22,9 @@
 def get_args():
     parser = argparse.ArgumentParser()
-    parser.add_argument("--task", type=str, default="HalfCheetah-v2")
+    parser.add_argument("--task", type=str, default="Hopper-v2")
```
why this change?

BFAnas replied:
Hopper is more challenging than HalfCheetah for offline RL in my experience

MischaPanch commented on examples/offline/d4rl_cql.py:
```diff
@@ -43,6 +43,7 @@ def get_args():
     parser.add_argument("--temperature", type=float, default=1.0)
     parser.add_argument("--cql-weight", type=float, default=1.0)
     parser.add_argument("--with-lagrange", type=bool, default=True)
+    parser.add_argument("--calibrated", type=bool, default=False)
```
Does it make sense to change to true? Should also add a description (copied from the docstring)

BFAnas replied:
OK!

MischaPanch commented on test/offline/test_cql.py:
```diff
@@ -185,14 +185,6 @@ def save_best_fn(policy):
     def stop_fn(mean_rewards):
         return mean_rewards >= args.reward_threshold
 
-    def watch():
```
Why this change? Was the function unused?

BFAnas replied:
Yes it was unused

MischaPanch commented on tianshou/policy/base.py:
```diff
@@ -259,6 +259,12 @@ def map_action_inverse(
                     act = (np.log(1.0 + act) - np.log(1.0 - act)) / 2.0  # type: ignore
         return act
 
+    def process_buffer(
```
Pls add documentation on when and how often this is supposed to be called. Also, why should it be possible to accept and return None?

BFAnas replied:
Because it is used in BaseTrainer init method where base trainer accepts None in buffer argument, but I guess I can add an if statement to rule out None.

MischaPanch commented on tianshou/policy/imitation/cql.py:
```diff
@@ -167,6 +173,35 @@ def calc_random_values(
         return random_value1 - random_log_prob1, random_value2 - random_log_prob2
 
+    def process_buffer(self, buffer: ReplayBuffer | None) -> ReplayBuffer | None:
+        if buffer is None:
```
When would that be the case? The whole logic of calibrated=True is not applied then. If there are such cases, one should either accommodate them, or raise an informative error if `self.calibrated=True`

MischaPanch commented on tianshou/policy/imitation/cql.py:
```diff
@@ -45,6 +45,9 @@ class CQLPolicy(SACPolicy):
     :param float alpha_max: upper bound for clipping cql_alpha. Default to 1e6.
     :param float clip_grad: clip_grad for updating critic network. Default to 1.0.
+    :param calibrated: calibrate Q-values as in CalQL paper arXiv:2303.05479.
+        Useful for offline to online, and also was observed to achieve better reults
```
```suggestion
        Useful for offline pre-training followed by online training, 
        and also was observed to achieve better results
```

MischaPanch commented on tianshou/trainer/base.py:
```diff
@@ -167,7 +167,7 @@ def __init__(
 
         self.policy = policy
-        self.buffer = buffer
+        self.buffer = policy.process_buffer(buffer)
```
```suggestion
        if buffer is not None:
             buffer = policy.process_buffer()
        self.buffer = buffer
```

Now that I think about it - does it make sense outside of offline learning?

MischaPanch replied:
If you agree, I'd address this in a follow-up. Was thinking of adding a `_post_init` method to the trainer

BFAnas replied:
> Now that I think about it - does it make sense outside of offline learning?

I can think of a case If somebody wants to start with a previously collected buffer and wants to preprocess it before starting learning.

MischaPanch commented:
Thanks for addressing the comments, @BFAnas ! It's close to merge, just some final things to address. Note that you can check the typing part of CI by calling `poe format` and `poe lint` locally, e.g., to catch mypy errors before pushing

MischaPanch commented:
I'll have a look

Anas BELFADIL ***@***.***> schrieb am Sa., 23. Sept. 2023,
11:34:

> @MischaPanch <https://github.com/MischaPanch> some mypy checks fail that
> are not in my part of the code, some are with some notes do you know about
> them?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/thu-ml/tianshou/pull/915#issuecomment-1732266838>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIOKQSWVKVGXEVXLA4VKMOLX32UL5ANCNFSM6AAAAAA3T2ANSA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 
appliedAI Initiative GmbH. All rights reserved. Strictly confidential.



DISCLAIMER.- All the information contained in this document / presentation 
is confidential and is protected by intellectual and industrial property 
laws. 

appliedAI Initiative GmbH is the legitimate owner or authorized 
user of the intellectual and industrial property rights contained in this 
document / presentation including brands, texts, images, designs, 
structures, color combinations, logos, animations, audio and video files , 
etc., unless expressly stated otherwise. Any form of disclosure, 
reproduction, copy or distribution, in whole or in part, is prohibited, and 
its content cannot be used for other purposes without the express 
authorization of

appliedAI Initiative GmbH. While appliedAI Initiative 
GmbH has made an effort to provide accurate information, errors and 
omissions can occur. Consequently, appliedAI Initiative GmbH does not 
accept any responsibility for any loss or damage that may, directly or 
indirectly, result from any advice, opinion, information, representation or 
omission, whether due to negligence or otherwise, to data or information 
that contained in this document / presentation

BFAnas commented:
> Thanks for addressing the comments, @BFAnas ! It's close to merge, just some final things to address. Note that you can check the typing part of CI by calling `poe format` and `poe lint && poe type-check` locally, e.g., to catch mypy errors before pushing

Thank you for the thorough reviews!

MischaPanch commented on tianshou/policy/base.py:
```diff
@@ -261,8 +261,12 @@ def map_action_inverse(
-        buffer: ReplayBuffer | None,
-    ) -> ReplayBuffer | None:
+        buffer: ReplayBuffer,
+    ) -> ReplayBuffer:
```
A more correct annotation would be

```python
TBuffer = TypeVar("TBuffer", bound=ReplayBuffer)

    ...
    def process_buffer(self, buffer: TBuffer) -> TBuffer:
       ...
```

You don't have to do it now, if you don't want to, but I think it'd be nice :)

MischaPanch commented:
I can't open a PR to your fork, could you change the settings? My commits are here:
https://github.com/aai-institute/tianshou/tree/review/calibrated-cql

BFAnas commented:
Hi @MischaPanch,
Sorry I was a little busy this week, I merged your changes now. I fixed some pep8 errors but there now the test are failing because of override "Module not found error", not sure if it's worth have it as a dependency. I think if we can do without it'll be better. 
Btw I don't find the "Allow anyone to contribute checkbox" in my fork settings.

MischaPanch commented:
@BFAnas Thanks! The overrides dependency will be added in #950 anyway, so might as well add it here. Pls run `poetry add overrides` and commit the pyproject and poetry.lock file (after a round of `poe format`).

In python 3.12 override will become part of the language, so might as well start using it now. For tianshou it's even particularly useful. After that, the MR is ready for merge, thank you a lot for the contribution!

BFAnas commented:
> @BFAnas Thanks! The overrides dependency will be added in #950 anyway, so might as well add it here. Pls run `poetry add overrides` and commit the pyproject and poetry.lock file (after a round of `poe format`).
> 
> In python 3.12 override will become part of the language, so might as well start using it now. For tianshou it's even particularly useful. After that, the MR is ready for merge, thank you a lot for the contribution!

Done!

MischaPanch commented on tianshou/policy/imitation/cql.py:
```diff
@@ -167,6 +177,23 @@ def calc_random_values(
+        """
+        if self.calibrated:
+            buffer = deepcopy(buffer)  # prevent mutation of original buffer
+            returns = calculate_returns_from_buffer(buffer, self.gamma)
```
```suggestion
            returns = calculate_returns_from_buffer(buffer, self._gamma)
```

MischaPanch commented on tianshou/policy/base.py:
```diff
@@ -259,6 +260,18 @@ def map_action_inverse(
+        **Note**: this will only be called once, when the trainer is initialized!
+        If the buffer is empty by then, there will be nothing to process.
+        This method is meant to be overridden by policies which will be trained
+        offline at some stage, e.g., in a pretraining step.
```
```suggestion
        offline at some stage, e.g., in a pre-training step.
```

MischaPanch commented:
@Trinkle23897 this one is ready to merge, but the GPU runner seems to have problems, and I can't click the merge button. Could you do that, or fix the GPU runner? Alternatively, you could change the repository settings such that I can merge without all checks being green

MischaPanch commented on tianshou/policy/base.py:
```diff
@@ -563,3 +581,52 @@ def _nstep_return(
+    done_field: str = "done",
+    reward_field: str = "rew",
+) -> np.ndarray:
+    """Calculate discounted returns from a replay buffer, taking into.
```
Docstring broken, but I can fix it in a subsequent PR. This one should be merged already

MischaPanch commented on examples/offline/utils.py:
```diff
@@ -1,5 +1,6 @@
 import d4rl
-import gymnasium as gym
+import d4rl.gym_mujoco
+import gym
```
That's a problem, we don't use gym anymore. I'm surprised it didn't show up in tests, it's unclear why this change was necessary at all.

@BFAnas could you pls fix this? Since you're already pushing then, pls also fix the comment above regarding the docstring

Trinkle23897 replied:
this is example, which isn't covered by test

MischaPanch commented on tianshou/data/buffer/her.py:
```diff
@@ -180,9 +182,10 @@ def rewrite_transitions(self, indices: np.ndarray) -> None:
 
         # Re-write meta
         assert isinstance(self._meta.obs, BatchProtocol)
+        # ep_obs = cast(BatchProtocol, ep_obs)
```
```suggestion
```

Trinkle23897 commented on tianshou/policy/imitation/cql.py:
```diff
@@ -167,6 +177,23 @@ def calc_random_values(
 
         return random_value1 - random_log_prob1, random_value2 - random_log_prob2
 
+    @override
```
override or overrides?

MischaPanch replied:
overrides is the package, the annotation is override (also in python 3.12), so it's correct

MischaPanch commented on test/policy/test_base.py:
```diff
@@ -0,0 +1,10 @@
+import numpy as np
```
I propose the tests to have the same structure and names as the src, makes finding the tests for a module very simple. This would be the start of it. WDYT?

Trinkle23897 commented on tianshou/policy/base.py:
```diff
@@ -563,3 +581,52 @@ def _nstep_return(
     return target_q.reshape(target_shape)
+
+
+# TODO: is there another method somewhere in tianshou to do this? Can _nstep_return be used?
```
We already have `compute_episodic_return` in BasePolicy
